src/flags.jl

```diff
@@ -0,0 +1,226 @@
+    0x0080 the last segment in the template
+    0x0100 secondary alignment
+    0x0200 not passing filters, such as platform/vendor quality controls
+    0x0400 PCR or optical duplicate
+    0x0800 supplementary alignment
+"""
+
+function flag end
+
+# Bitwise flags (or FLAG).
+for (name, bits, doc) in [
+        (:PAIRED,        UInt16(0x001), "the read is paired in sequencing, no matter whether it is mapped in a pair"),
+        (:PROPER_PAIR,   UInt16(0x002), "the read is mapped in a proper pair"                                       ),
+        (:UNMAP,         UInt16(0x004), "the read itself is unmapped; conflictive with FLAG_PROPER_PAIR"            ),
+        (:MUNMAP,        UInt16(0x008), "the mate is unmapped"                                                      ),
+        (:REVERSE,       UInt16(0x010), "the read is mapped to the reverse strand"                                  ),
+        (:MREVERSE,      UInt16(0x020), "the mate is mapped to the reverse strand"                                  ),
+        (:READ1,         UInt16(0x040), "this is read1"                                                             ),
+        (:READ2,         UInt16(0x080), "this is read2"                                                             ),
+        (:SECONDARY,     UInt16(0x100), "not primary alignment"                                                     ),
+        (:QCFAIL,        UInt16(0x200), "QC failure"                                                                ),
+        (:DUP,           UInt16(0x400), "optical or PCR duplicate"                                                  ),
+        (:SUPPLEMENTARY, UInt16(0x800), "supplementary alignment"                                                   ),]
+    @assert bits isa UInt16 "The bits must be of type UInt16."
+    sym = Symbol("FLAG_", name)
+    docstring = """    $sym
+    SAM/BAM flag: $doc
+
+    See also: [`flag`](@ref)
+    """
+    @eval begin
+        @doc $(docstring) const $(sym) = $(bits)
+    end
+end
+
+"""
+    ispaired(record::XAMRecord)::Bool
+
+Query whether the `record`'s template has multiple segments in sequencing.
+"""
+function ispaired(record::XAMRecord)::Bool
+    return flag(record) & FLAG_PAIRED == FLAG_PAIRED
+end
+
+"""
+    isproperpair(record::XAMRecord)::Bool
+
+Query whether each segment of the `record`'s template properly aligned according to the aligner.
+"""
+function isproperpair(record::XAMRecord)::Bool
+    return flag(record) & PROPER_PAIR == PROPER_PAIR
+end
+
+"""
+    isunmapped(record::XAMRecord)::Bool
+
+Query whether the `record` is unmapped.
+"""
+function isunmapped(record::XAMRecord)::Bool
+    return flag(record) & FLAG_UNMAP == FLAG_UNMAP
+end
+
+"""
+    ismapped(record::XAMRecord)::Bool
+
+Query whether the `record` is mapped.
+"""
+function ismapped(record::XAMRecord)::Bool
+    # return flag(record) & FLAG_UNMAP == 0
+    return isfilled(record) && (flag(record) & FLAG_UNMAP == 0)
```

Comment:
This seems to be a special case, in that it checks if the record is filled. I'm not sure this is a good idea:

1. Why do we even have unfilled records? We could just initialize all records as filled with empty data, then we could delete this check from the entire codebase. As I see it, it's an annoying and probably unnecessary edge case we could eliminate.
2. In this case, an unfilled record will neither be mapped or unmapped, which could be very confusing.
3. Why is it only this predicate function that checks if the record i filled?
4. IF this must take the filled status into account, it should be documented IMO

Reply:
I don't know why you'd want a record without data, either.
It may be the time to use StringViews in XAM; with hindsight, would you do it any other way?

Reply:
Having looked at this a bit closer, I see that removing `isfilled` is not straightforward as the behaviour of `ismapped` is woven into the tests. 
Since redesigning `record` initialisation is beyond the scope of this release, I'll leave this as is for now.
However, redesigning the `record` initialisation should be tackled in the next release.